src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleEqualityTests.cs

```diff
@@ -0,0 +1,2957 @@
+        _ = !t1; // error 4
+    }
+}";
+            var comp = CreateStandardCompilation(source, references: s_valueTupleRefs);
+            comp.VerifyDiagnostics(
+                // (7,13): error CS0019: Operator '+' cannot be applied to operands of type '(int, int)' and '(int, int)'
+                //         _ = t1 + t1; // error 1
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "t1 + t1").WithArguments("+", "(int, int)", "(int, int)").WithLocation(7, 13),
+                // (8,13): error CS0019: Operator '>' cannot be applied to operands of type '(int, int)' and '(int, int)'
+                //         _ = t1 > t1; // error 2
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "t1 > t1").WithArguments(">", "(int, int)", "(int, int)").WithLocation(8, 13),
+                // (9,13): error CS0019: Operator '>=' cannot be applied to operands of type '(int, int)' and '(int, int)'
+                //         _ = t1 >= t1; // error 3
+                Diagnostic(ErrorCode.ERR_BadBinaryOps, "t1 >= t1").WithArguments(">=", "(int, int)", "(int, int)").WithLocation(9, 13),
+                // (10,13): error CS0023: Operator '!' cannot be applied to operand of type '(int, int)'
+                //         _ = !t1; // error 4
+                Diagnostic(ErrorCode.ERR_BadUnaryOp, "!t1").WithArguments("!", "(int, int)").WithLocation(10, 13)
+                );
+        }
+
+        [Fact]
+        public void TestTypelessTuples()
+        {
+            var source = @"
+class C
+{
+    static void Main()
+    {
+        string s = null;
+        System.Console.Write((s, null) == (null, s));
+    }
+}";
+            var comp = CreateStandardCompilation(source, references: s_valueTupleRefs, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "True");
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var tuple1 = tree.GetCompilationUnitRoot().DescendantNodes().OfType<TupleExpressionSyntax>().ElementAt(0);
+            Assert.Equal("(s, null)", tuple1.ToString());
+            var tupleType1 = model.GetTypeInfo(tuple1);
+            Assert.Null(tupleType1.Type);
+            Assert.Equal("(System.String, System.String)", tupleType1.ConvertedType.ToTestDisplayString());
+
+            var tuple2 = tree.GetCompilationUnitRoot().DescendantNodes().OfType<TupleExpressionSyntax>().ElementAt(1);
+            Assert.Equal("(null, s)", tuple2.ToString());
+            var tupleType2 = model.GetTypeInfo(tuple2);
+            Assert.Null(tupleType2.Type);
+            Assert.Equal("(System.String, System.String)", tupleType2.ConvertedType.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void TestSimpleTypelessTupleAndTupleType()
+        {
+            var source = @"
+class C
+{
+    static void Main()
+    {
+        var t1 = (1, 2L);
+        System.Console.Write(t1 == (1L, 2));
```

Comment:
Why are these considered typeless tuples?

Reply:
Those tuples indeed have types. I'll fix the test name.

---
In reply to: [168941330](https://github.com/dotnet/roslyn/pull/23356#discussion_r168941330) [](ancestors = 168941330)